rust/lance/src/dataset.rs

```diff
@@ -4123,4 +4126,72 @@ mod tests {
             dataset.latest_version_id().await.unwrap()
         );
     }
+
+    #[tokio::test]
+    async fn test_nested_field_meta_roundtrips() {
+        fn item_field_with_meta() -> Field {
+            let mut item_field = Field::new("item", DataType::Utf8, true);
+            item_field.set_metadata(HashMap::from([(
+                COMPRESSION_META_KEY.to_string(),
+                "zstd".to_string(),
+            )]));
+            item_field
+        }
+
+        // generate a list<utf8> array with metadata
+        fn generate_list_string(batch_size: usize) -> GenericListArray<i32> {
+            let mut item_field = item_field_with_meta();
+
+            let mut builder = arrow::array::ListBuilder::new(arrow::array::StringBuilder::new())
+                .with_field(item_field);
+            while builder.len() < batch_size {
+                // foo_0, foo_1, foo_2, ...
+                let names: Vec<Option<String>> =
+                    (0..20).map(|i| Some(format!("foo_{}", i))).collect();
+                builder.append_value(names);
+            }
+            let names_array = builder.finish();
+            names_array
+        }
+
+        // write a dataset with a nested field with metadata
+        let item_field = item_field_with_meta();
+        let names_field = Field::new("names", DataType::List(Arc::new(item_field)), false);
+        let schema = Arc::new(arrow::datatypes::Schema::new(vec![names_field]));
+
+        let names = generate_list_string(10);
+        let schema_clone = schema.clone();
+        let mut options = RecordBatchOptions::default();
+        // options.match_field_names = false;
+        let batch =
+            RecordBatch::try_new_with_options(schema, vec![Arc::new(names)], &options).unwrap();
+
+        let batches = RecordBatchIterator::new([Ok(batch)], schema_clone);
+
+        let write_params = WriteParams {
+            mode: Overwrite,
+            data_storage_version: Option::from(LanceFileVersion::V2_1),
+            ..Default::default()
+        };
+
+        let test_dir = tempdir().unwrap();
+        let data_path = test_dir.path().to_str().unwrap();
+
+        Dataset::write(batches, data_path, Some(write_params))
+            .await
+            .unwrap();
+
+        // read the dataset
+        let dataset = Dataset::open(data_path).await.unwrap();
+
+        let scanner = dataset.scan();
+        let result = scanner.try_into_stream().await.unwrap();
+        let mut batch_stream = result.map(|b| b);
+        let mut batches = Vec::new();
+        while let Some(batch) = batch_stream.next().await {
+            assert!(batch.is_ok());
```

Comment:
The error occurs at this point, but for some reason, the exact error message is not reported by `cargo test` (it crashes without providing useful information). However, if the test case is placed in `lance/examples/write_read_ds.rs`, the error message described in [issue #2947](https://github.com/lancedb/lance/issues/2947) will be displayed.